src/back/hlsl/mod.rs

```diff
@@ -6,29 +6,99 @@ Backend for [HLSL][hlsl] (High-Level Shading Language).
-The inverse transposition happens when we call `[0]` to get the zeroth column vector.
+WGSL's ["Internal Layout of Values"][ilov] rules specify how each WGSL
+type should be stored in `uniform` and `storage` buffers. The HLSL we
+generate must access values in that form, even when it is not what
+HLSL would use normally.
 
-Because all of our matrices are implicitly transposed, we flip arguments to `mul`. `mat * vec`
-becomes `vec * mat`, etc. This acts as the inverse transpose making the results identical.
+The rules described here only apply to WGSL `uniform` variables. WGSL
+`storage` buffers are translated as HLSL `ByteAddressBuffers`, for
+which we generate `Load` and `Store` method calls with explicit byte
+offsets. WGSL pipeline inputs must be scalars or vectors; they cannot
+be matrices, which is where the interesting problems arise.
 
-The only time we don't get this implicit transposition is when reading matrices from Uniforms/Push Constants.
-To deal with this, we add `row_major` to all declarations of matrices in Uniforms/Push Constants.
+## Row- and column-major ordering for matrices
 
-Finally because all of our matrices are transposed, if you use `mat3x4`, it'll become `float3x4` in HLSL
-(HLSL has inverted col/row notation).
+WGSL specifies that matrices in uniform buffers are stored in
+column-major order. This matches HLSL's default, so one might expect
+things to be straightforward. Unfortunately, WGSL and HLSL disagree on
+what indexing a matrix means: in WGSL, `m[i]` retrieves the `i`'th
+*column* of `m`, whereas in HLSL it retrieves the `i`'th *row*. We
+want to avoid translating `m[i]` into some complicated reassembly of a
+vector from individually fetched components, so this is a problem.
 
-# Matrix struct member of the form `matCx2` Note
+However, with a bit of trickery, it is possible to use HLSL's `m[i]`
+as the translation of WGSL's `m[i]`:
 
-Struct member matrices of the form `matCx2` are translated to a sequence of C `vec2`s due to
-differences in alignment between WGSL and HLSL for uniform buffers.
+- We declare all matrices in uniform buffers in HLSL with the
+  `row_major` qualifier, and transpose the row and column counts: a
+  WGSL `mat3x4<f32>`, say, becomes an HLSL `row_major float4x3`. Since
+  the HLSL type is the transpose of how WebGPU directs the user to
+  store the data, HLSL will load all matrices transposed.
 
-Accesses to these matrices are handled by injected functions.
+- Since matrices are transposed, an HLSL indexing expression retrieves
+  the "columns" of the intended WGSL value, as desired.
+
+- For vector-matrix multiplication, since `mul(transpose(m), v)` is
+  equivalent to `mul(v, m)` (note the reversal of the arguments), and
+  `mul(v, transpose(m))` is equivalent to `mul(m, v)`, we can
+  translate WGSL `m * v` and `v * m` to HLSL by simply reversing the
+  arguments to `mul`.
+
+## Padding in two-row matrices
+
+An HLSL `row_major floatRx2` matrix has padding between its rows that
```

Comment:
Should be `row_major float2xC`.

Reply:
Are you sure? It's stored in row major order, so it's a series of rows. There is padding, so the padding must be between one row and the next. In order for there to be padding not present in WGSL, each row must be two values long. Hence, the matrix must have two columns.

Reply:
I think it's the WGSL type on the next line that's wrong: an HLSL `row_major floatRx2` matrix would represent a WGSL `mat2xC` matrix, not a `matCx2`.

Reply:
I was confused - the original `matCx2` was correct. If what's stored in memory is some series of two-element vectors (and HLSL and WGSL disagree about the padding between them), then since WGSL uses a column-major storage order, those two-element vectors must each be a WGSL `vec2<f32>`, of which there are `C`. That's `matCx2`.

Reply:
Ah yes it should be right in that order (`row_major floatRx2`). I got confused about the naming since I was expecting `C` to appear somewhere (akin to `matCx2`).

I think a shared value between the two notations would be helpful here (something like `row_major floatKx2`, `matKx2`).

What do you think?

Reply:
excellent suggestion